### PR TITLE
Precompile modexp test cases

### PR DESCRIPTION
**Pull request by jochem-brouwer**
## 🗒️ Description
This adds a few test cases for the ModExp precompile introduced in Byzantium.

Tested with: `evm version 1.13.4-stable-3f907d6a, solc version 0.8.23+commit.f704f362.Linux.gpp`

## 🔗 Related Issues
Part of these test cases where consensus bugs found in EthereumJS, see the issue: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3168 and the PR: https://github.com/ethereumjs/ethereumjs-monorepo/pull/3169

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](../docs/) directory.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
I need some help starting with the Tox validation. Is there a way to auto-fix my files?

**Review by danceratopz — COMMENTED** [review]
Hey @jochem-brouwer, thanks for contributing! 

Despite draft status, I think it's worth mentioning that a test can't call `state_test` multiple times within one test function execution (the framework should really test for this and raise an error if this condition is violated - I just created #365 to address this).

Instead, you can parametrize the test using pytest with the test vectors defined in data. Then the test function will get executed multiple times, once for each test vector. See https://github.com/jochem-brouwer/execution-spec-tests/pull/1.

**Comment by jochem-brouwer** [comment]
@danceratopz Thanks for the follow-up PR :smile: 

Although I have not added all cases from the EIP test file, I think this is ready for review. If you want, I can also add the remaining tests from the EIP.

**Comment by jochem-brouwer** [comment]
Note: I can not set labels for the PR, and I am assuming that if this gets merged, it will be merged via a squash-and-merge. I have not yet ticked those boxes.

**Comment by jochem-brouwer** [comment]
I have updated and also added the 2 remaining cases in the EIP.

File: tests/byzantium/precompiles/test_05_modexp.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        ),
+        pytest.param(
+            [
+                # Note: the only case which goes out-of-gas, and this is also raw input
+                # which is thus not fed into create_modexp_tx-_data
+                """0000000000000000000000000000000000000000000000000000000000000000"""
+                + """0000000000000000000000000000000000000000000000000000000000000020"""
+                + """ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"""
+                + """fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe"""
+                + """fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffd"""
```
Comment (jochem-brouwer) [inline tests/byzantium/precompiles/test_05_modexp.py:86]: This line and the one above fail to pass the spell check... should I add this to the whitelist? (Why are the other hex strings not failing the whitelist though? )

**Comment by jochem-brouwer** [comment]
Happy new years everyone :smile: 

Could someone take a look at this PR? I think it is ready-to-merge, except this spell issue (see my comment here: https://github.com/ethereum/execution-spec-tests/pull/364#discussion_r1435705248)

CC @danceratopz @marioevz

**Comment by danceratopz** [comment]
> Happy new years everyone 😄
> 
> Could someone take a look at this PR? I think it is ready-to-merge, except this spell issue (see my comment here: [#364 (comment)](https://github.com/ethereum/execution-spec-tests/pull/364#discussion_r1435705248))
> 
> CC @danceratopz @marioevz

Happy New Years @jochem-brouwer! Yup, will do!

**Review by danceratopz — COMMENTED** [review]
Hey @jochem-brouwer sorry for the delay! 

This is looking really good, below are a few suggestions. 

After you've looked at these comments, I might make a suggestion how to change the test IDs to make them more compact and re-usable in the console. Currently, they're a little bit messy to copy-paste and use them in the console due to the single quotes and spaces:
```
tests/byzantium/precompiles/test_05_modexp.py::test_modexp[fork_Byzantium-input_['', '', '02']-output_['0x01', '0x01']]
tests/byzantium/precompiles/test_05_modexp.py::test_modexp[fork_Byzantium-input_['', '', '0002']-output_['0x01', '0x0001']]
tests/byzantium/precompiles/test_05_modexp.py::test_modexp[fork_Byzantium-input_['00', '00', '02']-output_['0x01', '0x01']]
```
This is very pytest specific though, so happy to take care of that, if that's ok.

The other suggestion I'd make would be to move the tests:
- From: `tests/byzantium/precompiles/test_05_modexp.py`
- To: `tests/byzantium/eip198_modexp_precompile/test_modexp.py`
as precompiles were introduced over several forks, for consistency in our repo, I think it's better to organize by the EIP that they were added. We can make discovering and executing precompiles more convenient by adding pytest marks to the tests (e.g., `@pytest.mark.precompile_test`, see `fill --markers`, if you you're interested, it's like tagging tests with metadata).

**Comment by jochem-brouwer** [comment]
Regarding the test names, do you mean something like this?: 460bf35834943154dc66935c3c38372aa92c11f2

(I am resolving the latest comment as we speak :smile: )

**Comment by jochem-brouwer** [comment]
I'm not sure, how do I run `evm_bytes_to_python`? I can run `fill`, or `order_fixtures`, but if I try `evm_bytes_to_python` then I get `command not found`. (I have updated my branch to latest `main`)

**Comment by jochem-brouwer** [comment]
If I rename to https://github.com/ethereum/execution-spec-tests/pull/364/commits/fe4c0a8a18a54a0e999e22d0dac247c78b83455c

`tests/byzantium/eip198_modexp_precompile/test_modexp.py` then I have to add

```python
REFERENCE_SPEC_GIT_PATH = "EIPS/eip-198.md"
REFERENCE_SPEC_VERSION = "9e393a79d9937f579acbdcb234a67869259d5a96"
```

To the test (which I did). However now the documentation fails to build and I don't know why (just checked, CI passes if I rename it back https://github.com/ethereum/execution-spec-tests/actions/runs/7480915178/job/20361369965?pr=364 ). How should I fix this? It seems it can't find some files

```
ERROR   -  mkdocstrings: tests.byzantium.eip198_modexp_precompile.test_modexp could not be found

Aborted with a BuildError!
ERROR   -  Error reading page 'tests/byzantium/eip198_modexp_precompile/test_modexp/index.md':
ERROR   -  Could not collect 'tests.byzantium.eip198_modexp_precompile.test_modexp'
```

But I don't see anything in the other EIP tests which gives me a hunch what the problem could be :thinking:

**Comment by jochem-brouwer** [comment]
Never mind, I forgot to rename the `__index__.py` to the right folder.

@danceratopz ready for re-review :smile:

File: tests/byzantium/eip198_modexp_precompile/test_modexp.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        data=data,
+        gas_limit=500000,
+        gas_price=10,
+        protected=True,
+    )
+    if output[0] != "0x00":
+        # Note: This account is only created if the CALL output is 1. Otherwise, it is not created.
+        contract_address = compute_create_address(account, tx.nonce)
+        post[contract_address] = Account(code=output[1])
+    post[account] = Account(storage=dict([("0x00", output[0])]))
```
Comment (danceratopz) [inline tests/byzantium/eip198_modexp_precompile/test_modexp.py:189]: Just wanted to point out this is a more compact way of defining a dict with one value:
```suggestion
    post[account] = Account(storage={"0x00": output[0]})
```
No need to apply this, I modified it in my PRs (https://github.com/jochem-brouwer/execution-spec-tests/pull/2 and https://github.com/jochem-brouwer/execution-spec-tests/pull/3).

**Review by danceratopz — COMMENTED** [review]
Amazing! This looks really nice! Thanks for all your work on this.

> Regarding the test names, do you mean something like this?: https://github.com/ethereum/execution-spec-tests/commit/460bf35834943154dc66935c3c38372aa92c11f2
Yes that's one approach (and I like it a lot), but I was originally hoping to use auto-generated names, but make them more compact. 

In the end I decided to refactor how we write the parameter values using dataclasses in the hopes of making the code more readable. I quite like how it turned out, what do you think?

I made two PRs one with manual ids only (closer to this PR as it currently stands) and one which generates ids from the parameters. We should pick the one we like most and merge that :smiley_cat: - let me know what you think.
cc @marioevz @spencer-tb please also say which of the following you like more... I'm really torn :laughing: And what do you think about the dataclass usage? I think the overhead makes sense for these tests.
- https://github.com/jochem-brouwer/execution-spec-tests/pull/2
- https://github.com/jochem-brouwer/execution-spec-tests/pull/3

**Comment by jochem-brouwer** [comment]
I personally like this one better https://github.com/jochem-brouwer/execution-spec-tests/pull/2 since it is more descriptive. (I had thought of creating those manually, but indeed creating those using autogen is better :smile: )

Let me know if you need anything from my side :smile: :+1:

**Comment by jochem-brouwer** [comment]
Also regarding the new classes: that definitely makes it more readable!

**Comment by spencer-tb** [comment]
Big fan of 2) for the maintainability :D - will be a nice structure to follow for similar tests in the future

File: tests/byzantium/eip198_modexp_precompile/test_modexp.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        (
+            ModExpInput(base="", exponent="", modulus="01"),
+            ExpectedOutput(call_return_code="0x01", returned_data="0x00"),
+        ),
+        (
+            ModExpInput(base="", exponent="", modulus="0001"),
+            ExpectedOutput(call_return_code="0x01", returned_data="0x0000"),
+        ),
+        # Test cases from EIP 198 (Note: the cases where the call goes out-of-gas and the
+        # final test case are not yet tested)
```
Comment (danceratopz) [inline tests/byzantium/eip198_modexp_precompile/test_modexp.py:141]: I all the test cases from the EIP have been implemented now?
```suggestion
        # Test cases from EIP 198.
```

**Review by danceratopz — APPROVED** [review]
Thanks again @jochem-brouwer! Nice edition.

Can you just check the comment below? Otherwise, LGTM!

**Comment by jochem-brouwer** [comment]
@danceratopz I have just manually checked, the EIP has 5 test cases. I manually also checked that the test cases from the EIP match the test cases as written down in this PR. So from my side also: go ahead and merge, thanks a lot :smile: :+1:

File: tests/byzantium/eip198_modexp_precompile/test_modexp.py (hunk 3 above)
Comment (jochem-brouwer) [reply tests/byzantium/eip198_modexp_precompile/test_modexp.py]: Yup :smile:

**Pull request merged**